javapoet/src/main/java/com/palantir/javapoet/TypeSpec.java

```diff
@@ -1007,7 +1081,7 @@ public TypeSpec build() {
                         kind.implicitTypeModifiers);
             }
 
-            boolean isAbstract = modifiers.contains(Modifier.ABSTRACT) || kind != Kind.CLASS;
+            boolean isAbstract = modifiers.contains(Modifier.ABSTRACT) || (kind != Kind.CLASS && kind != Kind.RECORD);
```

Comment:
We can use a switch here to make this more clear

```suggestion
            boolean isAbstract = switch (kind) {
                case CLASS -> modifiers.contains(Modifier.ABSTRACT);
                case RECORD, ENUM, ANNOTATION -> false;
                case INTERFACE -> true;
            };
```